### PR TITLE
Implement P2440R1 ranges::iota, ranges::shift_left, ranges::shift_right

### LINKED ISSUE #2537: P2440R1 `ranges::iota`, `ranges::shift_left`, `ranges::shift_right`
[P2440R1](https://wg21.link/P2440R1) `ranges::iota`, `ranges::shift_left`, `ranges::shift_right`

New feature-test macro:
```cpp
#define __cpp_lib_ranges_iota 202202L
```

This also **increases** the value of an existing feature-test macro:
```cpp
#define __cpp_lib_shift 202202L
```

### PR DESCRIPTION
**Pull request by MitalAshok**
Fixes #2537

### CODE REVIEW

File: tests/std/tests/P0769R2_shift_left_shift_right/test.cpp [hunk 1]
```diff
@@ -7,6 +7,12 @@
 #include <iterator>
 #include <list>
 #include <vector>
+#include <version>
+
+#if __cpp_lib_shift >= 202202L
+// Test ranges::shift_left, ranges::shift_right from P2440R1
+#define TEST_RANGES
+#endif
```
Comment (fsb4000) [inline tests/std/tests/P0769R2_shift_left_shift_right/test.cpp:15]: I think we comment out `#endif`

So it should be 

`#endif // __cpp_lib_shift >= 202202L`

File: tests/std/tests/P0769R2_shift_left_shift_right/test.cpp [hunk 2]
```diff
@@ -106,12 +112,44 @@ void test_case_shift_left(const ptrdiff_t tmpSize) {
...
+            test_iota(first, last, pos_to_shift + 1, tmpSize);
+        }
+
+        {
+            fill_iota(tmp, tmpSize);
+            auto [ first, last ] = ranges::shift_left(tmp, pos_to_shift);
+            assert(first == tmp.begin());
+            test_iota(first, last, pos_to_shift + 1, tmpSize);
+        }
+#endif
```
Comment (fsb4000) [inline tests/std/tests/P0769R2_shift_left_shift_right/test.cpp:130]: Same here:

`#endif // TEST_RANGES`

and below.

File: stl/inc/numeric [hunk 3]
```diff
@@ -14,6 +14,11 @@
 #include <limits>
 #endif // _HAS_CXX17
 
+#ifdef __cpp_lib_concepts
+#include <algorithm>
```
Comment (MitalAshok) [inline stl/inc/numeric:18]: Including `<algorithm>` might be a bit heavy. We only need the definition of `out_value_result`, so maybe that could be extracted into a separate file?

**Comment by fsb4000** [comment]
Also add to first comment `Fixes #2537` so that issue will be closed after they will merge the PR.

[GitHub's close/fix/resolve syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

File: stl/inc/yvals_core.h [hunk 4]
```diff
@@ -1332,9 +1332,14 @@
 
 #define __cpp_lib_remove_cvref            201711L
 #define __cpp_lib_semaphore               201907L
-#define __cpp_lib_shift                   201806L
```
Comment (fsb4000) [inline stl/inc/yvals_core.h:1335]: In this file you should also add a comment about implementing the feature.

See https://github.com/microsoft/STL/blob/7b2c54be012f998736ebb6b72c7804a3f6c5f8e5/stl/inc/yvals_core.h#L282-L303

File: stl/inc/numeric (hunk 3 above)
Comment (frederick-vs-ja) [reply stl/inc/numeric:18]: I think `out_value_result` should be moved to `<xutility>`, which is included by both `<algorithm>` and `<numeric>` in MSVC STL. Note that `in_out_result` is already there.

File: stl/inc/numeric [hunk 5]
```diff
@@ -14,6 +14,11 @@
 #include <limits>
 #endif // _HAS_CXX17
 
+#ifdef __cpp_lib_concepts
+#include <algorithm>
+#include <iterator>
```
Comment (frederick-vs-ja) [inline stl/inc/numeric:19]: Not needed. The concepts used here are defined in `<xutility>`.

File: stl/inc/numeric [hunk 6]
```diff
@@ -523,6 +523,36 @@ _CONSTEXPR20 void iota(_FwdIt _First, _FwdIt _Last, _Ty _Val) {
...
+    constexpr _It _Iota_unchecked(_It _First, const _Se& _Last, _Ty& _Val) {
+        for (; _First != _Last; ++_First, (void) ++_Val) {
+            *_First = const_cast<const _Ty&>(_Val);
+        }
+        return _STD move(_First);
+    }
+
+    template <input_or_output_iterator _It, sentinel_for<_It> _Se, weakly_incrementable _Ty>
+        requires indirectly_writable<_It, const _Ty&>
+    constexpr iota_result<_It, _Ty> iota(_It _First, _Se _Last, _Ty _Val) {
```
Comment (miscco) [inline stl/inc/numeric:541]: Ditto niebloid

**Review by miscco — CHANGES_REQUESTED** [review]
Welcome to the STL and thanks for the contribution.

This PR has already a lot of things in place. That is awesome for a first contribution!

Unfortunately, we have to completely rework the algorithms themselfs to provide the guarantees from the standard.

If you have any further questions on how to do this ask away.

File: tests/std/tests/P0769R2_shift_left_shift_right/test.cpp [hunk 7]
```diff
@@ -7,6 +7,16 @@
 #include <iterator>
 #include <list>
 #include <vector>
+#include <version>
```
[the remark this answers is not in the document]
Comment (AlexGuteniev) [reply tests/std/tests/P0769R2_shift_left_shift_right/test.cpp:10]: I think we may consider reverse: adding all tests here. This will avoid adding too many small tests and unnecessarily increase test run time.

File: stl/inc/algorithm [hunk 8]
```diff
@@ -5118,6 +5118,155 @@ _FwdIt shift_right(_ExPo&&, _FwdIt _First, _FwdIt _Last, _Iter_diff_t<_FwdIt> _P
     // not parallelized as benchmarks show it isn't worth it
     return _STD shift_right(_First, _Last, _Pos_to_shift);
 }
+
+#if _HAS_CXX23 && defined(__cpp_lib_concepts)
+namespace ranges {
+    template <class _It, class _Se>
+    constexpr _It _Shift_left_unchecked(_It _First, const _Se _Last, _Iter_diff_t<_It> _Pos_to_shift) {
```
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/inc/algorithm:5125]: Note that there are several occurrences of `_Iter_diff_t` that should be changed.

File: stl/inc/algorithm [hunk 9]
```diff
@@ -5118,6 +5118,155 @@ _FwdIt shift_right(_ExPo&&, _FwdIt _First, _FwdIt _Last, _Iter_diff_t<_FwdIt> _P
...
+        _Seek_wrapped(_Result, _Shift_left_unchecked(_Get_unwrapped(_First), _Get_unwrapped(_Last), _Pos_to_shift));
+        return {_First, _Result};
+    }
+
+    template <forward_range _Rng>
+        requires permutable<iterator_t<_Rng>>
+    constexpr borrowed_subrange_t<_Rng> shift_left(_Rng&& _Range, range_difference_t<_Rng> _Pos_to_shift) {
+        auto _First  = _RANGES begin(_Range);
+        auto _Result = _First;
+        _Seek_wrapped(_Result, _Shift_left_unchecked(_STD move(_Get_unwrapped(_Result)), _Uend(_Range), _Pos_to_shift));
```
Comment (StephanTLavavej) [inline stl/inc/algorithm:5163]: `_STD move(_Get_unwrapped(` appears to be unintentional - there are no other occurrences, and `_Get_unwrapped()` already returns a prvalue.

File: stl/inc/algorithm [hunk 10]
```diff
@@ -5118,6 +5118,155 @@ _FwdIt shift_right(_ExPo&&, _FwdIt _First, _FwdIt _Last, _Iter_diff_t<_FwdIt> _P
...
+        return {_First, _Result};
+    }
+
+    template <class _It, typename _Se>
```
Comment (StephanTLavavej) [inline stl/inc/algorithm:5167]: Style: We follow an ancient convention of `class` instead of `typename` for template parameters in product code (although I personally prefer `typename`).

File: stl/inc/algorithm [hunk 11]
```diff
@@ -5118,6 +5118,155 @@ _FwdIt shift_right(_ExPo&&, _FwdIt _First, _FwdIt _Last, _Iter_diff_t<_FwdIt> _P
...
+                auto _Size = _ULast - _UFirst;
+                if (_Pos_to_shift >= _Size) {
+                    _Seek_wrapped(_First, _UEnd_at);
+                    return {_First, _First};
+                }
+                _RANGES advance(_UEnd_at, -_Pos_to_shift);
+            } else {
+                for (; 0 < _Pos_to_shift; --_Pos_to_shift) {
+                    if (_UEnd_at == _UFirst) {
+                        _Seek_wrapped(_First, decltype(_UFirst)(_ULast));
```
Comment (StephanTLavavej) [inline stl/inc/algorithm:5190]: `decltype(_UFirst)(_ULast)` is a functional-style cast (like `int(x)`), which we conventionally avoid in favor of C++ casts. The rationale is that functional-style casts have C semantics (e.g. they are willing to reinterpret pointers to integers) and are extremely difficult to search for.

Comment (StephanTLavavej) [inline stl/inc/algorithm:5190]: Instead of `decltype(_UFirst)` (which contains `const`), could `_Unwrapped` be used here?

File: stl/inc/algorithm [hunk 12]
```diff
@@ -5118,6 +5118,155 @@ _FwdIt shift_right(_ExPo&&, _FwdIt _First, _FwdIt _Last, _Iter_diff_t<_FwdIt> _P
...
+                        return {_First, _First};
+                    }
+                    --_UEnd_at;
+                }
+            }
+
+            _Seek_wrapped(_First, _RANGES _Move_backward_common(_UFirst, _UEnd_at, _ULast));
+            return {_First, _RANGES next(_First, _Last)};
+        } else if constexpr (sized_sentinel_for<_Se, _It> && bidirectional_iterator<_Se>) {
```
Comment (StephanTLavavej) [inline stl/inc/algorithm:5199]: Is `bidirectional_iterator<_Se>` intended, or should this be `bidirectional_iterator<_It>`?

File: stl/inc/algorithm [hunk 13]
```diff
@@ -5118,6 +5118,155 @@ _FwdIt shift_right(_ExPo&&, _FwdIt _First, _FwdIt _Last, _Iter_diff_t<_FwdIt> _P
     // not parallelized as benchmarks show it isn't worth it
     return _STD shift_right(_First, _Last, _Pos_to_shift);
 }
+
+#if _HAS_CXX23 && defined(__cpp_lib_concepts)
+namespace ranges {
+    template <class _It, class _Se>
+    constexpr _It _Shift_left_unchecked(_It _First, const _Se _Last, _Iter_diff_t<_It> _Pos_to_shift) {
+        if (_Pos_to_shift <= 0) {
```
Comment (StephanTLavavej) [inline stl/inc/algorithm:5126]: I see that both `shift_left` and `shift_right` are special-casing `_Pos_to_shift <= 0`, but the Standardese says "*Preconditions:* `n >= 0` is `true`." Thus, I believe that only `== 0` should be special-cased, and we should `_STL_ASSERT` that the value is non-negative.

File: tests/std/tests/P0769R2_shift_left_shift_right/test.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply tests/std/tests/P0769R2_shift_left_shift_right/test.cpp:10]: @AlexGuteniev Thanks for keeping test run time in mind - while it's true that new test directories do increase the total test run time, the added cost is small in comparison to the total cost. I'd generally say that adding a small number of test directories (1 to 3) for a significant new feature is fine. (It's more reasonable for very tiny bugfixes to add very tiny test cases to existing test directories.)

While it would be possible to fuse everything into this test, I think I prefer @MitalAshok's current organization here - the reason is that the new test directories follow the "instantiator" pattern of many existing ranges tests. (For example, I was able to diff the `generate` and `iota` tests which were structurally very similar, and that made things easier to verify and understand.) I think that fusing the tests together would obscure this pattern, and also be somewhat of a distraction from what `P0769R2_shift_left_shift_right` is currently testing.

@miscco Conversely, while it's generally a good idea to keep new test machinery together, I think there may be a compelling argument for what @MitalAshok is doing here - this test `P0769R2_shift_left_shift_right` is focused on runtime behavior (whereas the others are more compile-time focused) and already contains significant setup machinery to test various size and shift values. If this were extracted into another directory or directories, that machinery would need to be replicated, which is somewhat burdensome for maintenance.

So, to summarize: Although it could be done differently, I believe that the current organization is reasonable. Please meow if you still disagree (I am not a ranges expert and you've written a lot more tests recently than I have). :smile_cat:

File: tests/std/tests/P0769R2_shift_left_shift_right/test.cpp [hunk 14]
```diff
@@ -7,6 +7,16 @@
...
 #include <list>
 #include <vector>
+#include <version>
+
+#if __cpp_lib_shift >= 202202L
+// Test ranges::shift_left, ranges::shift_right from P2440R1
+//#define TEST_RANGES
+/*
+ * Disabled until `forward_list::_Unchecked_end` is fixed to return the same type as `forward_list::end()._Unwrapped()`
+ * to work with `_Uend` in <xutility>
```
Comment (StephanTLavavej) [inline tests/std/tests/P0769R2_shift_left_shift_right/test.cpp:17]: Is this something that can be fixed while preserving ABI? Or does it indicate an overly strong assumption in our unwrapping machinery (e.g. similar to #2591)?

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

File: stl/inc/algorithm [hunk 15]
```diff
@@ -5118,6 +5118,155 @@ _FwdIt shift_right(_ExPo&&, _FwdIt _First, _FwdIt _Last, _Iter_diff_t<_FwdIt> _P
...
+    template <class _It, class _Se>
+    constexpr _It _Shift_left_unchecked(_It _First, const _Se _Last, _Iter_diff_t<_It> _Pos_to_shift) {
+        if (_Pos_to_shift <= 0) {
+            _RANGES advance(_First, _Last);
+            return _First;
+        }
+
+        auto _Start_at = _First;
+
+        if constexpr (sized_sentinel_for<_Se, _It>) {
```
Comment (CaseyCarter) [inline stl/inc/algorithm:5133]: These two branches look a lot like:
```c++
if (_RANGES advance(_Start_at, _Pos_to_shift, _Last) > 0) {
  return _First;
}
```
to me.

File: stl/inc/algorithm [hunk 16]
```diff
@@ -5118,6 +5118,155 @@ _FwdIt shift_right(_ExPo&&, _FwdIt _First, _FwdIt _Last, _Iter_diff_t<_FwdIt> _P
...
+#if _HAS_CXX23 && defined(__cpp_lib_concepts)
+namespace ranges {
+    template <class _It, class _Se>
+    constexpr _It _Shift_left_unchecked(_It _First, const _Se _Last, _Iter_diff_t<_It> _Pos_to_shift) {
+        if (_Pos_to_shift <= 0) {
+            _RANGES advance(_First, _Last);
+            return _First;
+        }
+
+        auto _Start_at = _First;
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/algorithm:5131]: For non-`sized_sentinel_for`, `distance` traverses the entire range, but we need not traverse the entire range to determine that `_Pos_to_shift` < `distance`. In fact, `if (_Pos_to_shift < distance(r)) { advance(_Start_at, _Pos_to_shift); _Move_unchecked(...);` does two full traversals where `advance(_Start_at, _Pos_to_shift, _Last); _Move_unchecked(...)` only needs one.

File: stl/inc/algorithm [hunk 17]
```diff
@@ -5118,6 +5118,155 @@ _FwdIt shift_right(_ExPo&&, _FwdIt _First, _FwdIt _Last, _Iter_diff_t<_FwdIt> _P
...
+    template <permutable _It, sentinel_for<_It> _Se>
+    constexpr subrange<_It> shift_right(_It _First, _Se _Last, _Iter_diff_t<_It> _Pos_to_shift) {
+        _Adl_verify_range(_First, _Last);
+        return _Shift_right(_First, _Last, _Pos_to_shift);
+    }
+
+    template <forward_range _Rng>
+        requires permutable<iterator_t<_Rng>>
+    constexpr borrowed_subrange_t<_Rng> shift_right(_Rng&& _Range, range_difference_t<_Rng> _Pos_to_shift) {
+        return _Shift_right(_RANGES begin(_Range), _RANGES end(_Range), _Pos_to_shift);
```
Comment (CaseyCarter) [inline stl/inc/algorithm:5266]: This could use `_Ubegin` and `_Uend` instead of `_RANGES begin` and `_RANGES end`.

File: stl/inc/algorithm [hunk 18]
```diff
@@ -5118,6 +5118,155 @@ _FwdIt shift_right(_ExPo&&, _FwdIt _First, _FwdIt _Last, _Iter_diff_t<_FwdIt> _P
...
+                    _RANGES iter_swap(_Mid, _Trail);
+                }
+            }
+        }
+    }
+
+    template <permutable _It, sentinel_for<_It> _Se>
+    constexpr subrange<_It> shift_right(_It _First, _Se _Last, _Iter_diff_t<_It> _Pos_to_shift) {
+        _Adl_verify_range(_First, _Last);
+        return _Shift_right(_First, _Last, _Pos_to_shift);
```
Comment (CaseyCarter) [inline stl/inc/algorithm:5260]: We can `_Get_unwrapped` both `_First` and `_Last` now that we've `_Adl_verify_range`d them.

**Review by CaseyCarter — CHANGES_REQUESTED** [review]
Partial review from today's Open Code Review.

File: tests/std/tests/P2440R1_ranges_alg_shift_left/test.cpp [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+STATIC_ASSERT(same_as<decltype(ranges::shift_left(borrowed<false>{}, 1)), ranges::dangling>);
+STATIC_ASSERT(same_as<decltype(ranges::shift_left(borrowed<true>{}, 1)), ranges::subrange<int*>>);
+
+
+struct instantiator {
+    static constexpr int_wrapper expected_result[2] = {12345, 3333};
+    static constexpr int_wrapper expected_io[5]     = {12345, 3333, 44, -1, -1};
+
+    template <ranges::forward_range Rng>
+        requires permutable<ranges::iterator_t<Rng>>
```
Comment (CaseyCarter) [inline tests/std/tests/P2440R1_ranges_alg_shift_left/test.cpp:37]: We have a few tests that apply constraints in the instantiator, we conventionally do so via `if constexpr (meow)` in the body of call instead of using concept overloading like this. The cost of overload resolution is probably dwarfed by the cost of instantiating the actual test code, so it's probably not important for test performance, but it would be nice to be consistent with the pattern we use elsewhere. (Also applies to the `shift_right` test.)

File: tests/std/tests/P0769R2_shift_left_shift_right/test.cpp (hunk 14 above)
Comment (CaseyCarter) [reply tests/std/tests/P0769R2_shift_left_shift_right/test.cpp:17]: The unwrapping machinery assumes that `_Get_unwrapped(ranges::end(r))` and `_Uend(ranges::end(r))` are equivalent expressions (`_Uend` actually asserts that they have the same type). This isn't true for `forward_list`: `_Unchecked_end()` yields a sentinel (for nice codegen properties) but `end()` has to return an iterator because the Standard says so.

I removed the `static_assert` from `_Uend`, and taught the range overload of `_Get_final_iterator` that `_Uend` for a `common_range` might not return an iterator, in which case it needs to `_Get_unwrapped(end())` instead of returning `_Unchecked_end()`. (I recall no other places that need similar cleanup.)

**Review by CaseyCarter — APPROVED** [review]

File: stl/inc/algorithm [hunk 20]
```diff
@@ -5035,6 +5046,238 @@ _FwdIt shift_right(_ExPo&&, _FwdIt _First, _FwdIt _Last, _Iter_diff_t<_FwdIt> _P
...
+        using _Not_quite_object::_Not_quite_object;
+
+        template <permutable _It, sentinel_for<_It> _Se>
+        constexpr subrange<_It> operator()(_It _First, const _Se _Last, iter_difference_t<_It> _Pos_to_shift) const {
+            _STL_ASSERT(_Pos_to_shift >= 0, "shift count must be non-negative (N4910 [alg.shift]/5)");
+
+            _Adl_verify_range(_First, _Last);
+            auto _UFirst = _Get_unwrapped(_First);
+            auto _ULast  = _Get_unwrapped(_Last);
+            auto _Size   = _Size_helper(_UFirst, _ULast);
```
Comment (miscco) [inline stl/inc/algorithm:5127]: I guess for debug performance it would be nice to early return here in the `_Pos_to_shift == 0` case, rather than calling `_Size_helper` etc

File: stl/inc/algorithm [hunk 21]
```diff
@@ -5035,6 +5046,238 @@ _FwdIt shift_right(_ExPo&&, _FwdIt _First, _FwdIt _Last, _Iter_diff_t<_FwdIt> _P
...
+            _STL_INTERNAL_STATIC_ASSERT(permutable<_It>);
+            _STL_INTERNAL_STATIC_ASSERT(sentinel_for<_Se, _It>);
+
+            if (_Pos_to_shift == 0) {
+                return {_First, _RANGES next(_First, _Last)};
+            }
+
+            constexpr bool _Is_sized = !same_as<_SizeTy, _Unsized>;
+
+            if constexpr (_Is_sized) {
```
Comment (miscco) [inline stl/inc/algorithm:5192]: I do not like how we have quite different implementations between `shift_left` and `shift_right`

Above we directly check whether something is sized in the call operator. This here seems more convoluted.

File: stl/inc/algorithm (hunk 20 above)
Comment (StephanTLavavej) [reply stl/inc/algorithm:5127]: That's not a pure win, though - it costs an extra branch. (Perhaps optimized away in release mode, but definitely not in debug mode.)

In general, when early-return cases aren't required for correctness, I think they're best avoided by default, unless/until we have evidence that such cases are especially common.

File: stl/inc/algorithm (hunk 21 above)
Comment (StephanTLavavej) [reply stl/inc/algorithm:5192]: `shift_right` is inherently a more complicated algorithm, if I understand it correctly. This centralizes the size handling, which seems reasonable given that the code below is so large. If you can find a way to improve this in a followup PR, I'd be interested to see it.

**Review by StephanTLavavej — APPROVED** [review]

**Review by CaseyCarter — APPROVED** [review]

File: stl/inc/algorithm (hunk 20 above)
Comment (miscco) [reply stl/inc/algorithm]: It is a pure win, the first thing we do inside the helper function is to check against 0.

This would just hoist that check before the `_Size` calculation

File: stl/inc/algorithm [hunk 22]
```diff
@@ -5035,6 +5046,238 @@ _FwdIt shift_right(_ExPo&&, _FwdIt _First, _FwdIt _Last, _Iter_diff_t<_FwdIt> _P
...
+            return _Result;
+        }
+
+        template <class _It, class _Se, class _SizeTy>
+        _NODISCARD static constexpr subrange<_It> _Shift_right_impl(
+            _It _First, _Se _Last, iter_difference_t<_It> _Pos_to_shift, const _SizeTy _Size) {
+            _STL_INTERNAL_STATIC_ASSERT(permutable<_It>);
+            _STL_INTERNAL_STATIC_ASSERT(sentinel_for<_Se, _It>);
+
+            if (_Pos_to_shift == 0) {
```
Comment (miscco) [inline stl/inc/algorithm:5186]: @StephanTLavavej nothing happens before this check so we can hoist it up

File: stl/inc/algorithm [hunk 23]
```diff
@@ -5035,6 +5046,238 @@ _FwdIt shift_right(_ExPo&&, _FwdIt _First, _FwdIt _Last, _Iter_diff_t<_FwdIt> _P
...
+                        return {_Buf, _Buf};
+                    }
+                }
+
+                if constexpr (_Is_sized) {
+                    if (_Size < 2 * _Pos_to_shift) {
+                        return {_Buf, _Move_n_helper(_First, _Size - _Pos_to_shift, _Buf)};
+                    }
+                }
+                auto _Lead = _Buf;
```
Comment (miscco) [inline stl/inc/algorithm:5236]: Could we add some lines between the non chained  controlflow

Comment (CaseyCarter) [reply stl/inc/algorithm:5236]: This is all one paragraph to me, with the goal of determining if `next(_Buf, _Pos_to_shift)` is valid and defining `_Lead` to that value if so, so it makes sense to me this way. I'll add a newline before the definition of `_Lead`.

File: stl/inc/algorithm (hunk 21 above)
Comment (CaseyCarter) [reply stl/inc/algorithm:5192]: Bidirectional `shift_right` is almost identical to `shift_left`. Forward `shift_right` is a completely different beast, almost a variation of `rotate` that throws away the contents of one of the subranges. I think this style - `if constexpr` dispatch between sized and unsized versions of the bounds-checking and setup, followed by the nested loops that perform the actual work - is overall more readable than the alternative of sized and unsized versions of the entire algorithm. I sometimes carry DRY to extremes and end up writing more code to avoid duplication, but that's not the case here.

File: stl/inc/algorithm (hunk 23 above)
Comment (StephanTLavavej) [reply stl/inc/algorithm:5234]: Sounds good to me. The scenario where our conventions insist on a newline is between `if`/`else if`/`else` chains: `if (is_cute) { } if (is_fluffy) { } if (is_kitten) { }`. Newlines in other situations are basically "author's preference". Personally, I like to have a newline between a condition and a variable, which is what Casey's adding here: `if (cond) { } Type var;`, since this is like "taking a breath" before starting something new. Whether there should be a newline between a variable and a following condition, as in `Type var; if (another_cond) { }` seems context-dependent to me (having a newline is a bit easier to read; omitting it signals that the variable is strongly associated with the following condition).

Thanks for always thinking about code clarity! :smile_cat:

File: stl/inc/algorithm [hunk 24]
```diff
@@ -5035,6 +5046,236 @@ _FwdIt shift_right(_ExPo&&, _FwdIt _First, _FwdIt _Last, _Iter_diff_t<_FwdIt> _P
...
+                    if (_RANGES advance(_Buf, _Pos_to_shift, _Last) != 0) {
+                        return {_Buf, _Buf};
+                    }
+                }
+
+                if constexpr (_Is_sized) {
+                    if (_Size < 2 * _Pos_to_shift) {
+                        return {_Buf, _Move_n_helper(_First, _Size - _Pos_to_shift, _Buf)};
+                    }
+                }
```
Comment (StephanTLavavej) [inline stl/inc/algorithm:5232]: No change requested: I observe that this `if constexpr (_Is_sized)` could be fused into the previous or the following `if constexpr (_Is_sized)` (and the following one comments about this `_Size < 2 * _Pos_to_shift` test). However, I see an argument for splitting them up this way, so that the previous and the following conditions are narrowly focused on advancing `_Buf` and `_Lead`, so this is fine as-is.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks and congratulations for implementing this significant C++23 feature! :tada: :heart_eyes_cat: :rocket:

This will ship in VS 2022 17.4 Preview 1.
